File: connect.c
/**
 * @author f4prime
 * @email zzz136454872@163.com
 * @aim a plain model for leetcode
 */

#include"ctools.h"
#include<stdio.h>
#include<stdlib.h>
#include<string.h>

//#define testMod
#ifdef testMod
void test() 
{

}
#endif

#define max(a,b) (((a)>(b))?(a):(b))
#define min(a,b) (((a)<(b))?(a):(b))

#ifndef testMod
//Definition for a Node.
struct Node {
    int val;
    struct Node *left;
    struct Node *right;
    struct Node *next;
};

struct Node * getNode(int val)
{
    struct Node* out=(struct Node*)malloc(sizeof(struct Node));
    out->val=val;
    out->left=NULL;
    out->right=NULL;
    out->next=NULL;
    return out;
}

struct Node* connect(struct Node* root) {
    struct Node* head=root;
    struct Node* prev,*nextHead;
    struct Node* now;
    while(head!=NULL)
    {
        now=head;
        prev=NULL;
        nextHead=NULL;
        while(now!=NULL)
        {
            printf("now: %d\n",now->val);
            if(now->left!=NULL)
            {
                if(nextHead==NULL)
                    nextHead=now->left;
                if(prev!=NULL)
                    prev->next=now->left;
                prev=now->left;
            }
            if(now->right!=NULL)
            {
                if(nextHead==NULL)
                    nextHead=now->right;
                if(prev!=NULL)
                    prev->next=now->right;
                prev=now->right;
            }
            now=now->next;
        }
        head=nextHead;
    }
    return root;
}
#endif

int main()
{
#ifdef testMod
    test();
#endif
#ifndef testMod
    struct Node* root=getNode(1);
    //root->left=getNode(2);
    root->right=getNode(2);
    //root->left->left=getNode(4);
    //root->left->right=getNode(5);
    //root->right->right=getNode(7);
    //root=connect(root);
    struct Node* tmp=root->right;
    printf("%x\n",tmp->next);
#endif
    return 0;
}
    

File: connect.py
from typing import *

class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        head=root
        while head!=None:
            nextHead=None
            pre=None
            while head!=None:
                if head.left!=None:
                    if nextHead==None:
                        nextHead=head.left
                    if pre!=None:
                        pre.next=head.left
                    pre=head.left
                if head.right!=None:
                    if nextHead==None:
                        nextHead=head.right
                    if pre!=None:
                        pre.next=head.right
                    pre=head.right
                head=head.next
            head=nextHead
        return root

sl=Solution()
root=Node(1)
root.left=Node(2)
root.left.left=Node(4)
root.left.right=Node(5)
root.right=Node(3)
root=sl.connect(root)
print(root.left.left.next.val)

